conda/gateways/connection/download.py

```diff
@@ -89,11 +92,22 @@ def download(url, target_full_path, md5sum=None, progress_update_callback=None):
                 log.debug("%s, trying again" % e)
             raise
 
-        actual_md5sum = digest_builder.hexdigest()
-        if md5sum and actual_md5sum != md5sum:
-            log.debug("MD5 sums mismatch for download: %s (%s != %s), "
-                      "trying again" % (url, digest_builder.hexdigest(), md5sum))
-            raise MD5MismatchError(url, target_full_path, md5sum, actual_md5sum)
+        if md5:
```

Comment:
Is it worthwhile to compute both sha256 and md5? My original thought was to only compute the most secure option available. I guess also computing md5 won't add much time, though.

Reply:
Using sha256 instead of md5 isn't really a valid security measure.  You need to couple sha256 with cryptographic signatures to actually have "security" in this context.  Without signatures, the best assurance we can have here is that the artifact wasn't inadvertently corrupted by erroneous HTTP over the wire.  It's false to think that md5 is any more "secure" in this use case than sha256.  (Probably why s3 uses md5 for etags, as I mentioned in https://github.com/conda/conda/pull/8651.)  So while it's not any more secure, you also incur the performance penalty of the heavier sha256.

All that said, we'll probably need to start using sha256 and incorporating that into any upcoming package signing implementation, so while we're really not adding any security now, it makes sense to prepare for the future.  I'll make the change.

Reply:
> Is it worthwhile to compute both sha256 and md5?

Computing both is wasteful because computing SHA256 usually isn't `x` times but rather `0.x` times slower than computing MD5. On my machine it's roughly 50 %.

> Using sha256 instead of md5 isn't really a valid security measure.

Sure.

> So while it's not any more secure, you also incur the performance penalty of the heavier sha256.
> \[...\]
> All that said, we'll probably need to  start using sha256 and incorporating that into any upcoming package signing implementation

For performance reasons you might want to consider SHA512 or SHA512/256 (supported from `openssl 1.1.1` onwards, available in Python via `cryptography >=2.5`). For me ( = not representative, of course), computing SHA512/256 is nearly as performant as MD5:
```
> podman run --rm -it continuumio/miniconda3 sh -lic '                                                                                                                                                                                                                      
conda create -qynopenssl openssl >/dev/null && conda activate openssl
openssl version
for hash in md5 sha256 sha512 sha512-256 ; do openssl speed -seconds 2 -bytes 1048576 -evp "${hash}" 2>/dev/null | tail -1 ; done
'
OpenSSL 1.1.1b  26 Feb 2019
md5             790102.02k
sha256          515857.24k
sha512          767557.63k
sha512-256      770179.07k
```

Reply:
cool!  thanks for the pointer.